Fixes for issues #1-3

robot256 opened the pull request:
Hope this helps.  It solves some things that were bothering me.  Love the mod, going to make another Youtube demonstration video soon!

robot256 committed:
Fixed Issue #1

Fixed spelling in debug messages, Issue #1 https://github.com/piero-la-lune/Factorio-Dispatcher/issues/1

robot256 committed:
Fix spelling

Fixed rest of Issue #1 https://github.com/piero-la-lune/Factorio-Dispatcher/issues/1

robot256 committed:
Fix issue #2

Fix issue #2 (https://github.com/piero-la-lune/Factorio-Dispatcher/issues/2):

Changed control.lua so that when trains are split, trains without any locomotives are removed from the dispatch and awaiting-dispatch lists and are set to manual mode.  This prevents unexpected behavior, such as "no path" messages being produced by uncoupled wagons.  It also prevents a manual train from being suddenly set to automatic when the player couples to wagons that were previously part of a dispatched train.

robot256 committed:
Fix Issue #3

Fix issue #3 (https://github.com/piero-la-lune/Factorio-Dispatcher/issues/3)

Changed the "dispatcher-debug dump" command to use OpteraLib's logger function.  The game data tables are formatted and written to factorio_current.log in the game directory for later vieweing.

robot256 committed:
Fixed debug prints

Debug prints go to console, debug dumps go to log file.

robot256 commented:
Adding a note that fixing this allows uncoupled wagons to be loaded by inserters while they wait.

robot256 committed:
Handle case when trains merge and new train has no schedule

Trains without a schedule will be removed from the dispatched and awaiting_dispatch tables.

Also removed checks for zero locomotives in the merging case because that won't ever happen.

robot256 commented:
Here are some more comments I'm saving here to investigate.  I'll reply again if I get to test any of them.

robot256's temporary notes:
1. Global serpent library might be able to be used simply by removing the `local serpent = require("serpent")` line, but it's not clear that it still exists in the game's `lualib` directory. _This worked, and I pushed a commit that restores the old dump feature and adds a new "dumplog" command to write to the log file._
2. Station names don't need to be updated every 300 ticks.  Instead, use the `on_entity_died`, `on_player_mined_entity`, `on_robot_mined_entity`, `script_raised_destroy`, and `on_entity_renamed` event to trigger updates of individual units.  _Added in a branch, seems to work great, so I merged it to my pull request!_
3. Station name table does not need to contain every single station, only the stations which are Dispatcher stations and those with valid dispatcher post-fixes.  The name can be checked for correct format every time a "built" or "renamed" event is triggered, and added or removed from the table then.  As it is, the station table gets unnecessarily large on maps with many train stops but few Dispatcher routes.  _Included in the event trigger update._
4. When using Creative Mod, sometimes stations would be created but never added to the dispatcher list.  I believe this will be fixed by handling the `script_raised_built` and `script_raised_revive` events just like `on_built_entity` and `on_robot_built_entity`.  _Added, seems to work but I never was able to reliably reproduce it so we'll have to see._
5. The `globals.awaiting_dispatch` and `globals.dispatched tables` do not reliably purge train ids when those trains are merged, split, or destroyed.  This is a potential memory leak when problematic merge/splits are automated and repeated thousands of times in one map.  It's not totally clear if events can be used to determine when a train id is no longer valid, so this may need to be periodically purged.  At the very least, when `train_created` gives a merged train, both `old_train_id`s can be purged.  _Got this working too!  It only does the purge on_configuration_change, though._

robot256 committed:
Changed logging again to remove dependency

Fixed reference to built-in serpent method (no require statement is needed at all).  Restored dispatcher-debug command to dump tables to console.  Added a new dispatcher-debug command to dump tables to log file.

robot256 committed:
Added train-stop detection to script_raised_built and script_raised_revive

robot256 committed:
Changed global.stations to be maintained by events rather than polling

Stations are now stored in the table as global.stations[backer_name][unit_number].
Stations are only stored in the table if they are train-stop-dispatcher entities or their name ends in ".<number>".
Events are used to determine when a station entry may need to be added, removed, or renamed.

robot256 committed:
Added code to remove old trains from global lists when they are merged/split

Also purges lists for nonexistent trains whenever configuration changes.

robot256 committed:
Fixes for 0.18

- Switched to using built-in LuaEntity::get_merged_signal method.
- Added "icon_mipmaps = 1" to the signal icon, so that it would RETURN ANY NONZERO VALUE on the signal.

robot256 committed:
More 0.18 updates

- Added icon_mipmaps=1 to the DispatcherTrainStop item.
- Updated info.json with bogus version number for testing.

piero-la-lune commented on info.json:
```diff
@@ -1,10 +1,10 @@
 {
   "name": "Dispatcher",
-  "version": "1.1.1",
+  "version": "1.1.201",
```
Any reason for 201 and not just 2?
Considering the large amount of changes, it might be best to go for 1.2.0

robot256 replied:
Just so that I would remember to go back and change it later.  Any number scheme is fine.

piero-la-lune replied:
Since it adds support for 0.18, I would go for 1.2.0

piero-la-lune commented on control.lua:
```diff
@@ -33,24 +35,118 @@ script.on_configuration_changed(function(data)
+function add_station(entity)
+  local name = entity.backer_name
+  local id = entity.unit_number
+  if entity.name == "train-stop-dispatcher" or name:match("%.[123456789]%d*$") then
```
Why use the second condition `name:match`? The first one should be enough, no?
Moreover, we might want to let the user decide of the naming convention, so we should not rely on that whenever possible.

robot256 replied:
The reason:  If you have a map with 1000 train stops, 1 of them is a dispatcher, and 5 of them are dispatcher-formatted, I do not want to store all 1000 names and entities in the same table.  That is bad for performance when searching the list, and also inconvenient when reading the dump output.  

If/when we allow changing the naming convention, we can chance the match string together with it.

piero-la-lune replied:
Ok I get it, I had not understood what you were doing. It is ok then, let's keep it that way.

piero-la-lune commented on control.lua:
```diff
@@ -33,24 +35,118 @@ script.on_configuration_changed(function(data)
+
+
+-- Scrub list of trains
+function scrub_trains()
```
Could you explain why this is needed? Edge case scenario I would presume?

robot256 replied:
I put it in the migration code because the old code sometimes left invalid trains in the queues and this will remove them during the upgrade process.  I don't know if there are any edge cases in the new code that will produce orphaned train entries also (for example, loading the map after uninstalling some modded trains, and an entire Awaiting Dispatch train disappears).

piero-la-lune replied:
That is what I hate the most with Factorio modding, having to maintain these tables of objects that may be changed by the game or mods in tons of way. Let's keep it that way for now.

piero-la-lune commented on control.lua:
```diff
@@ -146,12 +264,13 @@ function tick()
       -- Get the dispatch signal at the dispatcher
-      local signal = get_signal(v.station, SIGNAL_DISPATCH)
+      local signal = v.station.get_merged_signal(SIGNAL_DISPATCH)
+      debug("Signal is "..tostring(signal))
```
Seems to me that this debug command will spam the logs (every tick) + no reference to the train id

robot256 replied:
Yes, I added it when debugging the signal/icon_mipmap issue and forgot to take it out.

piero-la-lune replied:
Ok

piero-la-lune commented on control.lua:
```diff
@@ -196,7 +315,11 @@ function tick()
+            debug("Train #", v.train.id, " can't find any enabled station '", name, "'")
           end
+        else
+          --debug("Train #", v.train.id, " can't find any station named '", name, "'")
```
Why commented? It seems to be a useful log

robot256 replied:
This will also spam the logs every tick, so I remembered to take it out.

piero-la-lune replied:
Ah ok

piero-la-lune commented on control.lua:
```diff
@@ -205,103 +328,46 @@ end
 
 
--- Update list of stations (because players can change station names)
-function update_list_stations()
```
I do not remember why I did not use the event `on_entity_renamed` when I wrote this function (but I must have had a good reason at the time). Anyway, if it works with the event, it is ok with me.

robot256 replied:
They are adding/fixing API features all the time so there are lots of mods with funny workarounds for things that are easier in the latest version.  I first thought to do this because I noticed the CPU time for Dispatcher pulsating as it redid the station list periodically.  I like the new code because it uses no CPU at all if you do not change your rail network stations.

robot256 replied:
> Since 0.18 is out, and you made the changes required for the mod to work with it, I guess you did not plan to make a version for 0.17 (this is fine with me)?

I would like to release one version as 0.17 before moving on to 0.18.  It will make one of my saves better behaved.  Only the changes to signals.lua and items.lua are related to 0.18.

piero-la-lune replied:
Yes, it is much better for CPU, more readable, etc. So keep it that way.

piero-la-lune commented on control.lua:
```diff
@@ -205,103 +328,46 @@ end
-  if red then
-    value = red.get_signal(signal)
+-- Count the number of locomotives in the train
+function count_locos(train)
```
You do not want to know the number of locomotives, but only if there are any. Best to rename the function `has_loco`, and change the return value to `true` or `false`.

robot256 replied:
You are right.  I copied this from another mod, fixed it in that mod, and forgot to fix it here.

piero-la-lune commented on control.lua:
```diff
@@ -324,26 +390,43 @@ function print_game(...)
   game.print(text)
 end
 
+function print_file(...)
```
Is it possible to merge this function with `print_game` ?

robot256 replied:
Sure, we can combine the similar logic so it is only written once.

piero-la-lune reviewed (changes requested):
Hi robot256,

Thanks a lot for your contribution. I made a few minor comments, otherwise it seems good to me. I still have to run it for further tests, in particular performance, before approving it.

Since 0.18 is out, and you made the changes required for the mod to work with it, I guess you did not plan to make a version for 0.17 (this is fine with me)?

robot256 committed:
Changes and bugfixes requested by piero-la-lune

- Removed debug messages spamming every tick when destination is unavailable.
- Consolidated table/string printing code.
- count_locos -> has_locos
- Updated changelog.

robot256 commented:
> Since 0.18 is out, and you made the changes required for the mod to work with it, I guess you did not plan to make a version for 0.17 (this is fine with me)?

I would like to release one version as 0.17 before moving on to 0.18.  It will make one of my saves better behaved.  Only the changes to signals.lua and items.lua are related to 0.18.

robot256 committed:
Fixed changelog formatting to work in-game.

piero-la-lune commented:
I though `get_merged_signal` was 0.18, but it is already in `0.17`, so ok.

I still have to check it in the game, but tell me when you want me to merge it in master and release the 1.1.2 version.

piero-la-lune commented:
Could you make a separate merge request for 0.18, with the changes in the changelog and the two other files?

piero-la-lune commented:
I think I will create a new 1.2.0 branch, until 0.18 is out

robot256 commented:
Okay, I will revert the 0.18 changes on my master branch.  Then you can merge the major changes for 0.17, and make a branch for 0.18 with them included.  I will clone the branch, and the new pull request there.

I just started experimenting with using Stringy Train Stops and Dispatcher together, so I might find some more bugs this morning.  But I think it's stable enough to release tonight either way.

robot256 committed:
Reverted 0.18 changes so this can be released for 0.17

robot256 committed:
Update info.json

robot256 commented:
Okay, that's all the changes needed to make a good 0.17 release.  I'll make the 0.18 pull request when you close this one :)

piero-la-lune commented:
It is merged into master, and will be released soon on Factorio mod portal.

For your information, I discovered a bug (that was here before your pull request I think) : when 2 trains were merged while awaiting dispatch, the new train was added to the `awaiting_dispatch` table and set to automatic mode. But because it was set to auto mode, the `train_changed_state` event was triggered, and the new train removed from `awaiting_dispatch` while keeping awaiting schedule => bug
Anyway, I fixed it the same way splitting was already implemented, and it seems to work fine.

piero-la-lune commented:
1.1.2 is out, thanks a lot for your contribution @robot256 !

robot256 commented:
> For your information, I discovered a bug (that was here before your pull request I think)

You are right!  I didn't alter those lines of code.  That bug might explain some weird behavior I see occasionally where a train runs off on its own.  

Thanks for welcoming my changes!

robot256 commented:
I have made the changes for 0.18 in a branch of my fork.  If you update your 1.2.0 branch to master, I will make a pull request to merge my branch into that branch.

piero-la-lune commented:
Done
